CPR-509 add support for metadata files

e-gris opened the pull request:
- Add support for traversing trees of file metadata
- tweaked logging so it prints the script information
- try to make variables and functions descriptive
- update to version 0.2.0

mwaggett commented on README.rst:
```diff
@@ -1,17 +1,26 @@
 
-./generate_directory_indexes.py --help                                                                                     ~/development/generate_directory_indexes
-usage: generate_directory_indexes.py [-h] [--noop] [--verbose] path
+generate_directory_indexes.py [-h] [--file-metadata FILE_METADATA]
```
`--help` for consistency with the rest of the optional args?

mwaggett commented on README.rst:
```diff
@@ -1,17 +1,26 @@
-  -h, --help     show this help message and exit
-  --noop         Only print files to be created without writing them to disk
-  --verbose, -v  Verbose output. Repeat (up to -vvv) for more verbosity
+  -h, --help            show this help message and exit
```
perhaps reverse the order of these, for consistency

mwaggett commented on README.rst:
```diff
@@ -1,17 +1,26 @@
+  -h, --help            show this help message and exit
+  --file-metadata FILE_METADATA, -f FILE_METADATA
+                        A file containing data describing the tree to index.
+  --metadata-delimiter METADATA_DELIMITER, -m METADATA_DELIMITER
```
What's the default?

mwaggett commented on README.rst:
```diff
@@ -1,17 +1,27 @@
 
-./generate_directory_indexes.py --help                                                                                     ~/development/generate_directory_indexes
-usage: generate_directory_indexes.py [-h] [--noop] [--verbose] path
+generate_directory_indexes.py [--file-metadata FILE_METADATA]
```
you lost the `[-h]` altogether

e-gris replied:
Oopsies.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -15,37 +22,68 @@
+        logging.basicConfig(level=logging.DEBUG, format=logger_format)
+    else:
+        logging.basicConfig(level=logging.ERROR, format=logger_format)
+
```
I think you're missing a `fi` and an `end` to the `def`.

e-gris replied:
Nope. This is Python. No 'end' statements.

mwaggett replied:
oh right. ugh. It's been a while since I python'd.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -15,37 +22,68 @@
+    # filesystem path in the index
 
-    # Remove the base path from the prefix to avoid putting the full filesystem path in the index
     path = '' if prefix == base_path else prefix.replace(base_path, '')
```
I know this wasn't part of your changeset, but couldn't this just be `path = prefix.replace(base_path, '')`?

e-gris replied:
Quite possibly. I'd have to tinker with it to see if that excludes some edge case that isn't obvious. Left alone for now.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -93,20 +136,18 @@ def index_file_name(prefix, order_by, reverse_order):
+def format_file_details(file_details):
     out = []
-    for k in directory_listing:
+    for k in file_details:
```
What is `k`?

e-gris replied:
i, j, & k are conventionally reserved generic index variables for tight loops. I left these here.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -115,47 +156,131 @@ def format_directory_listing(directory_listing):
+    if last_modified == None:
+        last_modified = os.path.getmtime(full_path)
+    if size == None:
+        size =  os.path.getsize(full_path)
```
extra space in there, it looks like

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -115,47 +156,131 @@ def format_directory_listing(directory_listing):
+    return epoch_time
+
+
+def is_an_index_file(file_name):
```
Does python support question marks in function defs?

e-gris replied:
It doesn't.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -115,47 +156,131 @@ def format_directory_listing(directory_listing):
+    if file_name in index_file_names:
+        return True
+
+    return False
```
couldn't you just `return file_name in index_file_names`?

e-gris replied:
Wrong direction. The intent here is to provide a condition for not including certain files in the index.html files, namely the index.html files themselves. Previously there was a regex for that that I replaced with this whitelist. 

IOW, does current file match a special file? If Yes, then skip it.

A refactoring in the future will combine the logic of generating the index file name with creating the skippable-list. However, after some thought, that felt outside the scope of this already change-heavy PR.

mwaggett replied:
I understand how you're using this functions, but I'm pretty sure, if you zoom out, this just says
```
if (true)
   return True
return False
```
unless I'm just wrong about what python's `<blah> in <blah_blah>` does..

e-gris replied:
file_name will rarely be in index_file_names. It will likely be something like foo-bar-1.0.rpm. This function will return False most of the time.

mwaggett replied:
yeah, but my point is that this could just be simplified.
so sometimes, `return (true)`, but usually `return (false)`.
Am I making sense?

e-gris replied:
I could use:

```python
return True if file_name in index_file_names else False
```
but I don't find that terribly readable.

e-gris replied:
Now I understand your original comment. I misread it originally. You're right, I could use:

```python
return file_name in index_file_names
```
It *is* more succinct but isn't as obvious IMO.

mwaggett replied:
yeah, that's what I meant. There's definitely a balance between readability and succinctness, so I'll let you make the call.

e-gris replied:
After thinking about it overnight your version, I think, is correct. There's another tweak that I want to try as well.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -115,47 +156,131 @@ def format_directory_listing(directory_listing):
+    if file_metadata != None:
+        file_details = parse_file_metadata(current_path, file_metadata)
+    else:
+        file_details = gather_file_details(current_path, contents)
```
Is the reason for this ordering having non-default behavior first? because to me, having `if ... != ...` feels weird, but I do understand your reasoning.

e-gris replied:
Just a general pattern that I use of handling exceptional/edge/non-default cases first. This calls attention to the fact that they are exceptional and helps with preventing bugs.

mwaggett reviewed (approved):
assuming you've tested this somehow? 👍

mwaggett commented:
should probably squash the commits before merge, but otherwise lgtm 😊

demophoon commented on generate_directory_indexes.py:
```diff
@@ -1,8 +1,15 @@
-#!/usr/bin/python
+#!/usr/bin/env python
```
We should use `"""` for comment blocks instead of `## `

demophoon commented on generate_directory_indexes.py:
```diff
@@ -15,37 +22,68 @@
 
-# use a templating library to turn a prefix and a list of contents into an HTML directory index
+
+####
```
Same `"""` comment block issue as above. We should also move this comment into the function.
```py
def render_index(prefix, order_by, contents, reverse_order, base_path):
    """
    Use a templating library to turn a prefix and a list of contents
    into an HTML directory index
    """
```

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
+
+
+def is_excluded_file(file_name):
+    # A list of file names to exclude from the generated index files.
```
Same `"""` comment issue as above.

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
-    contents = os.listdir(path)
-    directory_listing = []
+
+def file_information(full_path, file, last_modified = None, size = None):
```
Don't insert spaces between the `=` in argument lists when setting argument defaults.

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
+def traverse_tree(base_path, current_path, file_metadata = None):
+    contents = os.listdir(current_path)
+
+    if file_metadata != None:
```
Nitpicky but you should use `is not` instead of `!=` when testing equality with `None`.

demophoon replied:
On second thought this could probably be simplified to just `if file_metadata`.

e-gris replied:
I like the explicit "is not"

demophoon replied:
Other than reading better it tests object identity instead of object equality. For things like `None`, `False`, and `True` we don't want to worry about what `__eq__` (or `==`) is going to do to test equality between the objects. With `is` and `is not` we actually test if the literal `None` we should expect is the exact same as the builtin `None` we are expecting.

However I believe that what we actually want is `if file_metadata:` instead. If `file_metadata` is ever `False` then this will equate to true resulting in the wrong thing happening.

e-gris replied:
Got it. Thanks.

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
-def validate_input(path):
-    if not os.path.isdir(path):
-        sys.exit('Error: {} is not a directory'.format(path))
+parse_arguments()
```
We should push these methods into some sort of boilerplate for python so that you can easily debug the file in a repl.

Try this instead:
```py
if __name__ == '__main__':
    parse_arguments()
    validate_input(configuration)
    traverse_tree(configuration.path, configuration.path, configuration.file_metadata)
```

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
+
+
+def validate_input(configuration):
+    global script
```
Since we are not mutating `script` in this function this is unnecessary.

demophoon commented on generate_directory_indexes.py:
```diff
@@ -93,20 +137,20 @@ def index_file_name(prefix, order_by, reverse_order):
+        if abs(a_number) < 1024.0:
+            return "%3.1f%s%s" % (a_number, unit, suffix)
+        a_number /= 1024.0
+    return "%.1f%s%s" % (a_number, 'Yi', suffix)
```
If possible we should be using `.format()` instead of `%` for formatting strings as `%` is deprecated.

demophoon commented on generate_directory_indexes.py:
```diff
@@ -15,37 +22,68 @@
-    logging.basicConfig(level=logging.INFO)
-else:
-    logging.basicConfig(level=logging.DEBUG)
+script = os.path.basename(sys.argv[0])
```
Is there any reason why `script` is in the top-level scope? I'm only seeing it be used in `validate_input`. We could push this down into that function's scope to declutter top-level scope as much as possible.

e-gris replied:
There *was* a reason for it but it's no longer applicable.

mwaggett commented on generate_directory_indexes.py:
```diff
@@ -22,7 +22,6 @@
 
 #sys.setrecursionlimit(2100000000)
```
probably want to remove this too?

demophoon commented on generate_directory_indexes.py:
```diff
@@ -115,47 +158,135 @@ def format_directory_listing(directory_listing):
+def file_information(full_path, file, last_modified=None, size=None):
+    if last_modified == None:
+        last_modified = os.path.getmtime(full_path)
+    if size == None:
```
~Nitpick but we should use `is` and not `==` when testing equality with `None`.~

Sorry. not this.. explanation comment coming soon.

demophoon replied:
See comment above for explanation of testing equality to `None`.

demophoon reviewed (approved):
This looks good to me

mwaggett commented:
@e-gris I will merge once you squash, unless you also want @danieldreier to take a look.

e-gris commented:
Let's not do this on a Friday afternoon. I'll squash again on Monday.

Commit:
CPR-509 add support for metadata files

  - Add support for traversing trees of file metadata
  - tweaked logging so it prints the script information
  - try to make variables and functions descriptive
  - update to version 0.2.0
  - changes from reviews
